src/Catel.MVVM/Catel.MVVM.Shared/MVVM/ViewModels/ViewModelManager.cs

```diff
@@ -173,18 +173,24 @@ public void UnregisterModel(IViewModel viewModel, object model)
 
             Log.Debug("Unregistering model '{0}' with view model '{1}' (id = '{2}')", modelTypeName, viewModelTypeName, viewModel.UniqueIdentifier);
 
+            var modelWasRemoved = false;
+
             lock (_viewModelModelsLock)
             {
-                if (!_viewModelModels.ContainsKey(viewModel.UniqueIdentifier))
+                if (_viewModelModels.TryGetValue(viewModel.UniqueIdentifier, out var models))
                 {
-                    _viewModelModels[viewModel.UniqueIdentifier] = new List<object>();
+                    models.Remove(model);
+                    modelWasRemoved = true;
                 }
+            }
 
-                var models = _viewModelModels[viewModel.UniqueIdentifier];
-                models.Remove(model);
+            if (modelWasRemoved)
+            {
+                Log.Debug("Unregistered model '{0}' with view model '{1}' (id = '{2}')", modelTypeName, viewModelTypeName, viewModel.UniqueIdentifier);
+                return;
```

Comment:
Try to keep the method flow alive (we don't know what we are adding to the end). I think using else is safer here.